engine-src/Game/LambdaHack/Client/UI/HandleHumanLocalM.hs

```diff
@@ -679,37 +679,44 @@ selectWithPointerHuman = do
 -- at terrain change or when walking over items.
 repeatHuman :: MonadClientUI m => Int -> m ()
 repeatHuman n = do
-  LastRecord _ seqPrevious k <- getsSession slastRecord
-  let macro = concat $ replicate n $ reverse seqPrevious
-  modifySession $ \sess -> sess {slastPlay = macro ++ slastPlay sess}
-  let slastRecord = LastRecord [] [] (if k == 0 then 0 else maxK)
-  modifySession $ \sess -> sess {slastRecord}
+  recording <- getsSession srecording
+  when (not recording) $ do 
+    macro <- getsSession smacroBuffer
+    let nmacro = concat $ replicate n $ reverse macro
+    modifySession $ \sess -> sess {slastPlay = nmacro ++ slastPlay sess}
 
-maxK :: Int
-maxK = 100
+-- * RepeatLast
+
+-- | Repeats last user's action.
+repeatLastHuman :: MonadClientUI m => m ()
+repeatLastHuman = do
+  lastAct <- getsSession slastAction
+  let cmd = fromMaybe [] $ pure <$> lastAct
+  modifySession $ \sess -> sess {slastPlay = cmd ++ slastPlay sess}
 
 -- * Record
 
 recordHuman :: MonadClientUI m => m ()
 recordHuman = do
-  lastPlayOld <- getsSession slastPlay
-  LastRecord _seqCurrent seqPrevious k <- getsSession slastRecord
-  case k of
-    0 -> do
-      let slastRecord = LastRecord [] [] maxK
-      modifySession $ \sess -> sess {slastRecord}
-      when (null lastPlayOld) $
-        -- Don't spam if recording is a part of playing back a macro.
-        promptAdd0 $ "Macro will be recorded for up to"
-                     <+> tshow maxK
-                     <+> "actions. Stop recording with the same key."
-    _ -> do
-      let slastRecord = LastRecord seqPrevious [] 0
-      modifySession $ \sess -> sess {slastRecord}
-      when (null lastPlayOld) $
-        -- Don't spam if recording is a part of playing back a macro.
-        promptAdd0 $ "Macro recording stopped after"
-                     <+> tshow (maxK - k - 1) <+> "actions."
+  isRecording <- getsSession srecording
+  if isRecording 
+  then do
+    -- Stop recording.
+    macro <- getsSession smacroBuffer
+    modifySession $ \sess ->
+      sess { srecording = False
+           , smacroBuffer = tail macro
```

Comment:
I think this explodes when the user runs macro `ComposeIfLocal Record Record` he previously defines in his config file.

Reply:
BTW, perhaps remove the spurious key when replaying the macro? then we don't need to read and write here.

Reply:
Thanks, didn't thought about that command. Indeed, this macro would crash a game.